Fixed - 8279 - [Enhancement] ListView ScrollTo with empty groups - Android

techduggu opened the pull request:
### Description of Change ###

<!-- Describe your changes here. If you're fixing a regression, please also include a link to the commit that first introduced this issue, if possible. -->
As requested in the ticket #8279, I've enhanced ListView ScrollTo method to scroll to a group with no child and also handled other use-cases discussed here: https://github.com/xamarin/Xamarin.Forms/issues/8279#issuecomment-547885319

### Issues Resolved ### 
<!-- Please use the format "fixes #xxxx" for each issue this PR addresses -->

- fixes #8279

### API Changes ###
My initial idea was to change the `GetGroupAndIndexOfItem()` method of `TemplatedItemsList.cs` to return corresponding group if there's no child or item found (we should be sending item index as zero instead of -1 to consider the groups with no child or no item found). The existing implementation returns `new Tuple<int, int>(groupIndex, -1);` which was the problem. But I noticed that `GetGroupAndIndexOfItem()` method is being used by other platform's ListViewRenderer and this approach also required further changes in the corresponding renderer to check for the group count. This isn't a good approach (and I don't want it to break for other platforms).

Hence, we need to handle the logic only in the corresponding `ListViewRenderer`. I've updated the following logic in `OnScrollToRequested` of `ListViewRenderer.cs` on Android to handle Scroll To use-cases as discussed above:
```
if (Element.IsGroupingEnabled)
{
	var results = templatedItems.GetGroupAndIndexOfItem(scrollArgs.Group, scrollArgs.Item);
	if (results.Item1 == -1 || (results.Item1 == -1 && results.Item2 == -1))
		return;

	int item = results.Item2 == -1 ? 0 : results.Item2;

	var group = templatedItems.GetGroup(results.Item1);
	if (group.Count == 0)
		cell = group.HeaderContent;
	else
		cell = group[item];

	position = templatedItems.GetGlobalIndexForGroup(group) + item + 1;
}
```

### Platforms Affected ### 
<!-- Please list all platforms affected by these changes -->
- Android

### Testing Procedure ###
<!-- Please list the steps that should be taken to properly test these changes on each relevant platform. If you were unable to test these changes yourself on any or all platforms, please let us know. Also, if you are able to attach a video of your test run, you will be our personal hero. -->

![8279_working](https://user-images.githubusercontent.com/17450941/67890984-bc479a80-fb77-11e9-975e-673d1c30df5c.gif)

- Just run `Issue8279.cs` from Control Gallery as this requires manual review for verification. I've included all the use-cases as discussed and demonstrated above in this test page.
Please Note: I don't have iOS setup so I'd need help of someone from Xamarin Team to review the above test on iOS (though I have not added anything to break other platforms).

### PR Checklist ###
<!-- To be completed by reviewers -->

- [x] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

jfversluis commented:
> as this requires manual review for verification

Why? I think it should be possible to write UI tests for this, right?

techduggu commented:
@jfversluis In an ideal case, it is possible to write UI tests and get confirmation by taking screenshots for this use-case but I thought that anyways requires a manual interaction (for verifying the screenshots). And I was more of thinking from a scroll position perspective (where I'd prefer it to verify manually whether it scrolls to the exact group/item or not). I also thought of making that automated but it requires computation for exact cell position of group/item (which I believe can be depicted from platform renderer only). All this approach was leading to quite complex state, hence I thought manual review would be better in this case.
But on a side note, I'm always ready to learn. If there's a better approach to automate this use-case, please suggest and I'm happy to implement it.
Thanks!

jfversluis commented:
Hm I think you might be overcomplicating this. In a UI test you can easily just let it tap the button and then detect if the group/item that it is supposed to scroll to is in view, as long as that group/item is uniquely identifiable.

Give your button and groups/items a value for the `AutomationId`, or for the groups/items you can simply use the text that is in there right now. Then in pseudo code it should be as simple as:

```
// To make sure the screen was loaded
RunningApp.WaitForElement("MyButtonAutomationId");
RunningApp.Tap("MyButtonAutomationId");

// This will fail if the list didn't scroll. If it did scroll, it will succeed
RunningApp.WaitForElement("AutomationIdOrTitleOfYourGroupOrItem");
```

Put this in a test method and create a method for each of the buttons at the top and we should have UI tests :)

Does that make sense?

techduggu commented:
@jfversluis That's as simple as awesome. I was in impression that `WaitForElement` method waits for an element to render on the page (initial load) but as you mentioned I just double checked its definition and it looks perfect for our use-case (to check the element in current view). Thanks for simplifying it (saves a lot of those cell-position calculation). I will add UI Tests for this.

techduggu commented:
@jfversluis I've added UI Tests as discussed. Please review.

jsuarezruiz commented:
Tested the changes on Android and looks nice. In iOS we would need to align these changes. Currently:

![issue8279](https://user-images.githubusercontent.com/6755973/68674857-fd1bb800-0556-11ea-849d-9df09f7f10b2.gif)

Could you check iOS?. I think you have not iOS setup, let me know if you need us to help you with iOS :)

techduggu commented:
@jsuarezruiz Thanks for looking into it. Yes, I don't have iOS setup and hence I've only implemented it for Android (also mentioned in the Title of this PR) and would need help from someone having iOS setup as I mentioned here:

> Please Note: I don't have iOS setup so I'd need help of someone from Xamarin Team to review the above test on iOS (though I have not added anything to break other platforms).

jsuarezruiz commented:
@techduggu Could you review the current conflict?. Thanks!.

techduggu commented:
@jsuarezruiz Done! Resolved the conflicts.

samhouts commented:
Seems like all tests were passing on the last run (2019-11-01) except for 

`Issue8145.Issue8145Shell Toolbar Disposed Exception`

which seems to be unrelated.

techduggu commented:
@samhouts Did you get the chance to view the test results? Just wanted to follow-up on this PR.

samhouts commented:
@techduggu it looks like there may have been a merge error. The build is failing. Are you able to look into that? Thanks!

techduggu commented:
@samhouts If I try to see the build error, it says 'Build not found'. Weird?

jfversluis commented:
Builds are not retained indefinitely. I have triggered a new one now!

techduggu commented:
Hi @jfversluis 
I think I have resolved the failing build with having an issue with the same name for UI test methods and constants defined (somehow this wasn't shown during my local build). Can you trigger a new build again as I don't see it automatically?

jfversluis commented:
Triggered it for you! Thank you for your efforts!

techduggu commented:
@jfversluis Thanks! Can you trigger it one last time? Finally, rebuilt it successfully. Sorry for any inconvenience caused.

jfversluis commented:
No worries, triggered again!

techduggu commented:
I can see now no more code related issue but it is failing at OSX and Nuget Phase. Can you assist here?

jfversluis commented:
I most certainly can! I see something about an invalid tag in the nuspec config. I think we updated some build bits while this PR is open. I think a rebase will help here. Are you familiar with that?

techduggu commented:
Thanks! Yup, I have taken latest from upstream/master and then merged it to this branch (basically rebased it on master). Can you trigger build again?

jfversluis commented:
Done and succeeded 🎉

rmarinho commented:
UItests on iOS that are failing are not related. But the test ScrollWithNoItemNoGroupTest is failing on API 19.

@hartez can you take a look at this PR ?!

rmarinho commented:
@techduggu  you mind rebase, and check UITest is failing on Android API19 ScrollWithNoItemNoGroupTest

techduggu commented:
Hi @rmarinho 
I've rebased it on master.

riverar commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8279.cs:
```diff
@@ -0,0 +1,226 @@
+	[Issue (IssueTracker.Github, 8279, "[Feature requested] ListView do not ScrollTo a group when there is no child of this group", PlatformAffected.Android)]
+    public class Issue8279 : TestContentPage
+	{
+        public static ListView List { get; set; }
```
Nit: Can you clean up the extra spaces here and throughout the file?

riverar commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8279.cs:
```diff
@@ -0,0 +1,226 @@
+		{
+
+		}
+		public Issue8279()
```
Nit: Add blank line after }

riverar commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8279.cs:
```diff
@@ -0,0 +1,226 @@
+        {
+#if APP
+			Data = new List<MyGroup>();
+            Data.Add(new MyGroup(){Headertitle = "Header 1"});
```
Nit: Spacing (see above comment)

riverar commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8279.cs:
```diff
@@ -0,0 +1,226 @@
+			Data.Last().Add(new MyData() { Title = "title 3b" });
+
+			List = new ListView();
+            List.HorizontalOptions = LayoutOptions.FillAndExpand;
```
Nit: Spacing

riverar commented:
@rmarinho Mind rerunning a build and pasting the test failure output?

samhouts commented:
Test failed on Android 27 with Fast Renderers: 

```
ScrollWithNoItemNoGroupTest_lg_nexus_5_4_4_4
System.Exception : Error while performing WaitForElement(Marked("Header 3"), "Timed out waiting for element...", 00:00:02, null, null)
----> System.TimeoutException : Timed out waiting for element...
```

riverar commented:
Last @samhouts spoke in chat, this was an unrelated Chrome EULA issue. Hoping we can move this forward.

techduggu commented:
@riverar Thanks for reviewing this and taking it forward.
Just wanted to update as I saw your review comments above on rebase, I am facing slowness with Xamarin.Forms project solution in my personal laptop (8 GB RAM and it just stucks with a heavy project solution build like this) and hence I am setting it up in another office machine (which has better configuration). I will try to rebase as soon as I set up my dev instance again.
Thanks!

samhouts commented:
@riverar That was actually a different PR....this one looks like it might be actually causing issues. 

This one crashed on iOS...
![image](https://user-images.githubusercontent.com/7827070/90567244-b9498500-e15e-11ea-8ac4-89eb4a546bdf.png)

Failed on Android....(and this is a new test from this PR, so definitely related)
![image](https://user-images.githubusercontent.com/7827070/90567312-cebeaf00-e15e-11ea-985e-4a85c73c905c.png)

And this one is an enigma... 🤷 
![image](https://user-images.githubusercontent.com/7827070/90567412-f6157c00-e15e-11ea-9c2c-762c12f6d077.png)

riverar commented:
@samhouts Need test names and the associated failures/stacks. Public does not have access to your test infrastructure and screenshots aren't informative.

riverar commented:
Here's one of the failing tests I can reproduce locally, looking into it.

```
Test FullName:	Xamarin.Forms.Core.Android.UITests.Xamarin.Forms.Controls.Issues.Issue8279.ScrollWithNoItemNoGroupTest
Test Source:	 : line -1
Test Outcome:	Failed
Test Duration:	0:00:00

Test Name:	ScrollWithNoItemNoGroupTest
Test Outcome:	Failed
Result StackTrace:	
at Xamarin.UITest.Utils.ErrorReporting.With[T](Func`1 func, Object[] args, String memberName)
   at Xamarin.UITest.Android.AndroidApp.WaitForElement(Func`2 query, String timeoutMessage, Nullable`1 timeout, Nullable`1 retryFrequency, Nullable`1 postTimeout)
   at Xamarin.Forms.Controls.ScreenshotConditionalApp.WaitForElement(Func`2 query, String timeoutMessage, Nullable`1 timeout, Nullable`1 retryFrequency, Nullable`1 postTimeout) in C:\Sources\Xamarin.Forms\Xamarin.Forms.Controls.Issues\Xamarin.Forms.Controls.Issues.Shared\TestPages\ScreenshotConditionalApp.cs:line 204
   at Xamarin.Forms.Controls.Issues.Issue8279.ScrollWithNoItemNoGroupTest() in C:\Sources\Xamarin.Forms\Xamarin.Forms.Controls.Issues\Xamarin.Forms.Controls.Issues.Shared\Issue8279.cs:line 145
--TimeoutException
   at Xamarin.UITest.Shared.WaitForHelper.WaitForAny[T](Func`1 query, String timeoutMessage, Nullable`1 timeout, Nullable`1 retryFrequency, Nullable`1 postTimeout)
   at Xamarin.UITest.Android.AndroidApp.<>c__DisplayClass80_0.<WaitForElement>b__0()
   at Xamarin.UITest.Utils.ErrorReporting.With[T](Func`1 func, Object[] args, String memberName)
Result Message:	
System.Exception : Error while performing WaitForElement(Marked("Header 3"), "Timed out waiting for element...", 00:00:02, null, null)
  ----> System.TimeoutException : Timed out waiting for element...
```

samhouts commented:
The iOS crash is on iOS 13 only. 

Issue6472.List View Does Not Throw Exception With Observable Collection

```
Issue8279.Scroll With No Item No Group Test
System.Exception : Error while performing WaitForElement(Marked("Header 3"), "Timed out waiting for element...", 00:00:02, null, null)
----> System.TimeoutException : Timed out waiting for element...
```

riverar commented:
```[Test]
public void ScrollWithNoItemNoGroupTest()
{
	RunningApp.WaitForElement(ScrollWithNoItemNoGroup);
	RunningApp.Tap(ScrollWithNoItemNoGroup);
	// This will pass if the list didn't scroll and remain on the same state
	RunningApp.WaitForElement(q => q.Marked("Header 3"), timeout: TimeSpan.FromSeconds(2));
}
```

It appears this test assumes a previous test has already completed and placed the list in a state where `Header 3` is selected. This will need to be fixed or removed.

One quick suggestion is to add a reset button to reset state, then add that to your tests:
```diff
+ const string Reset = "Reset";

// ...

+ var resetButton = new Button()
+ {
+    Text = "Reset",
+    AutomationId = Reset,
+    Command = new Command(()=> List.ScrollTo(null, firstGroup, ScrollToPosition.Center, true))
+ };

// ...

Content = new StackLayout () {
	VerticalOptions = LayoutOptions.StartAndExpand,
	HorizontalOptions = LayoutOptions.FillAndExpand,
-	Children = { button1, button2, button3, button4, button5, List },
+	Children = { resetButton, button1, button2, button3, button4, button5, List },
};

[Test]
public void ScrollWithNoItemNoGroupTest()
{
+	RunningApp.WaitForElement(Reset);
+	RunningApp.Tap(Reset);
	RunningApp.WaitForElement(ScrollWithNoItemNoGroup);
	RunningApp.Tap(ScrollWithNoItemNoGroup);
	// This will pass if the list didn't scroll and remain on the same state
-	RunningApp.WaitForElement(q => q.Marked("Header 3"), timeout: TimeSpan.FromSeconds(2));
+	RunningApp.WaitForElement(q => q.Marked("Header 1"), timeout: TimeSpan.FromSeconds(2));
}
```

cc: @techduggu

riverar commented:
@techduggu Any luck over there?

techduggu commented:
@riverar: Good news is I have setup a new instance of X.F solution in my office laptop and it's performing better than my personal laptop. I will try to fix the above mentioned changes first thing on Sunday morning along with rebase. Just occupied with office work and Saturday event: https://bdotnet.in/dotnet-dev-summit-2020/

riverar commented:
@techduggu great, thanks! I'll be here Sunday too if you run into any issues. Or if you need a VM set up for remote dev work. 🤖

techduggu commented:
@riverar I have added the above mentioned changes and also rebased. Can you trigger the build (as I see it didn't trigger automatically for all checks e.g. UI Tests).

Regarding space and style nits, I don't find it in my local (refer below screenshot). Are these spaces auto-added by Github preview pane? I saw @samhouts once asking the same on discord about automatically fixing such auto-added space nits.

![image](https://user-images.githubusercontent.com/17450941/91654807-a07b9080-eac9-11ea-9adc-b16993636e3f.png)

techduggu commented:
> @techduggu great, thanks! I'll be here Sunday too if you run into any issues. Or if you need a VM set up for remote dev work. 🤖

One quick question here, Does Xamarin Team provides such flexibility of providing VM in such cases? I meant when someone doesn't have a laptop with good configuration to build/work/contribute to Xamarin.Forms.

riverar commented:
/azp run

riverar commented:
@techduggu I'm a Windows MVP, not Microsoft FTE, so can't trigger builds. I'll ping some folks in Discord.

Regarding VMs: I was just offering a personal VM space. I suspect Microsoft could light up CodeSpaces for official Xamarin Forms development to make contribution much easier.

PureWeen commented:
@techduggu can you rebase against 5.0.0 and fix the conflict?

techduggu commented:
@PureWeen I saw we recently merged '5.0.0' into main, Should I still rebase against 5.0.0?
https://github.com/xamarin/Xamarin.Forms/commit/fce7c48f378d5f3578b70738d7f60e78dda0c93a

PureWeen commented:
@techduggu I think so

I changed the base to 5.0.0 and you'll see some extra files are coming in there

nickrandolph commented:
Unfortunately there are still issues with scrolling - these were probably there before this PR but I'll document them anyhow. 

If you follow the process of tapping Reset and then followed by each test scenario, they all work. However, if instead you scroll to the bottom of the list, and then click the test scenario, each of the scenarios that includes a Group only scrolls up so that part of the group is showing. I assume that this is not desirable and that the whole group heading should be visible

**With Item, With Group**
![Scroll_WithItem_WithGroup_FromBottom](https://user-images.githubusercontent.com/1614057/91919764-63521100-ed0a-11ea-8927-333e89f4e3fa.png)

**No Item, Empty Group**
![Scroll_NoItem_EmptyGroup_FromBottom](https://user-images.githubusercontent.com/1614057/91919767-65b46b00-ed0a-11ea-8bae-7c2bac3e1a09.png)

**No Item, With Group**
![Scroll_NoItem_WithGroup_FromBottom](https://user-images.githubusercontent.com/1614057/91919771-6816c500-ed0a-11ea-851a-708fe1035fee.png)

riverar commented:
Hey @nickrandolph! Yup, I believe this is pre-existing behavior, scrolling is wonky in general. `Xamarin.Forms.ScrollToPosition` plays a part in this, with its vague "scroll until it's visible" enum, which is satisfied with one row of pixels.

nickrandolph commented on Xamarin.Forms.Platform.Android/Renderers/ListViewRenderer.cs:
```diff
@@ -284,13 +284,18 @@ void OnScrollToRequested(object sender, ScrollToRequestedEventArgs e)
 			{
 				var results = templatedItems.GetGroupAndIndexOfItem(scrollArgs.Group, scrollArgs.Item);
-				if (results.Item1 == -1 || results.Item2 == -1)
+				if (results.Item1 == -1 || (results.Item1 == -1 && results.Item2 == -1))
```
Should this be
if (results.Item1 == -1 || (results.Item1 **!=** -1 && results.Item2 == -1))
The way it is currently, the second condition is irrelevant

techduggu replied:
@nickrandolph The original condition is correct as we need to return empty if either group alone is provided as null or both group and item parameters provided as null.
For e.g. (consider results.Item1 is indexOfGroup and results.Item2 is indexOfItem):
Scroll with no item but group will have values for indexOfGroup = 2 and indexOfItem = -1.
Scroll with no item no group will have values for indexOfGroup = -1

The above condition functionally correct for all given use-cases but I agree that it may be confusing for readability purpose due to 'Or' operator usage instead of if-else statements. The if-else version of same code will look like:
```
if (results.Item1 == -1)
 return;
else (results.Item1 == -1 && results.Item2 == -1)
 return;
```

Hence, please let me know if I should change it to if-else for better visibility.

nickrandolph replied:
else condition will never be hit, since it relies on the same condition as the if clause
from your explanation, you don't need the else clause

nickrandolph reviewed (changes requested):
My comment regarding the condition needs to be addressed before this is merged

nickrandolph commented on Xamarin.Forms.Platform.Android/Renderers/ListViewRenderer.cs:
```diff
@@ -284,13 +284,18 @@ void OnScrollToRequested(object sender, ScrollToRequestedEventArgs e)
 			if (Element.IsGroupingEnabled)
 			{
 				var results = templatedItems.GetGroupAndIndexOfItem(scrollArgs.Group, scrollArgs.Item);
```
Small thing but when we get the results from the GetGroupAndIndexOfItem can we assign Item1 and Item2 to local variables that have a meaning eg indexOfGroup and indexOfItem

techduggu replied:
@nickrandolph : Thanks for this suggestion. I added meaningful local variables.

nickrandolph commented on Xamarin.Forms.Platform.Android/Renderers/ListViewRenderer.cs:
```diff
@@ -280,13 +280,21 @@ void OnScrollToRequested(object sender, ScrollToRequestedEventArgs e)
+					cell = group[item];
 
-				position = templatedItems.GetGlobalIndexForGroup(group) + results.Item2 + 1;
+				position = templatedItems.GetGlobalIndexForGroup(group) + item + 1;
```
Can we provide a small comment explaining this line? eg why are we adding 1

nickrandolph commented on Xamarin.Forms.Platform.Android/Renderers/ListViewRenderer.cs:
```diff
@@ -280,13 +280,21 @@ void OnScrollToRequested(object sender, ScrollToRequestedEventArgs e)
 
-				var group = templatedItems.GetGroup(results.Item1);
-				cell = group[results.Item2];
+				int item = indexOfItem == -1 ? 0 : indexOfItem;
```
Can we change the variable "item" to something more meaningful. It appears to be an item index?

nickrandolph commented:
It would appear that there are additional files listed as changed now. Does this branch need a rebase?

nickrandolph commented:
@techduggu I think the behaviour of this PR is correct, so it's close to being able to be merged. A couple of outstanding things:
- Fix the condition I mentioned above (i.e. if (results.Item1 == -1 || (results.Item1 == -1 && results.Item2 == -1)) by removing the second part of the condition.
- Update the variable "item" and "position" to be more meaningful, and provide comment on the +1
- See if a rebase is required to remove changes to files that aren't really part of this PR.

techduggu commented:
@nickrandolph Thanks for reviewing and summarizing the points. I have incorporated the review comments and committed the files.
Regarding providing the comment on +1, As this was already existing code, I tried debugging again to understand and see that in the case of Grouping enabled, the original implementation is adding +1 for the scroll position to make it visible till the item. I can also see an existing comment at Line 300 i.e. Android offsets position of cells when using header. Please let me know if my understanding is correct.

Regarding the rebase, I already rebased to '5.0.0' as Shane mentioned and it doesn't show any further new changes for me. Am I missing anything?

![image](https://user-images.githubusercontent.com/17450941/92215220-13ae4980-eeb2-11ea-9d37-da5679a51f28.png)

nickrandolph reviewed (approved):
This PR is looking nice and neat. Thanks @techduggu for putting the effort in to rebase and clean up this PR. @samhouts / @jfversluis can we move this forward now?
